### ISSUE TITLE
SqlPackage Publish fails in some conditions with "ISqlModelElement: Identifier part cannot be null"

### ISSUE THREAD

**Issue by IVNSTN** [microsoft/DacFx#417]
- SqlPackage or DacFx Version: 162.2.111.2
- .NET Framework (Windows-only) or .NET Core:
- Environment (local platform and source/target platforms):
  - Teamcity agent: Windows Server 2012 R2
  - Microsoft SQL Server 2019 (RTM-CU25) - 15.0.4355.3 (X64)
  - Compatibility Level: 150

**Steps to Reproduce:**

In PR build of TSQL projects we run tSQLt tests in DB clone. For that we create DB clone on the fly and deploy corresponding dacpac with tests into this clone. This scenario works like a charm in all projects except one. The project is one of our In-Memory DBs; for the rest of In-Memory DBs described scenario works like a charm. The problematic DB can be deployed on "normal" DB (not clone) with no issues; sqlpackage params are the same, dacpac is exactly the same.

The problem can be reproduced on this project at any time, it is stable. But I have no glue what could be the cause of the problem.

SqlPackage params:
<pre>
& 'C:\Program Files\Microsoft SQL Server\160\dac\bin\sqlpackage.exe' /Action:Publish /SourceFile:'C:\BuildAgent\temp\buildTmp\output_build\post_trade_test.dacpac' /TargetConnectionString:"Data Source=our_server;Initial Catalog='_tsqlt_temp_clone__post_trade';Connect Timeout=60;Integrated Security=True;TrustServerCertificate=True;Persist Security Info=False;Pooling=False;MultipleActiveResultSets=False;Encrypt=False" 
/p:ExcludeObjectType=Logins /p:ExcludeObjectType=RoleMembership /p:ExcludeObjectType=Users /p:ExcludeObjectType=Permissions 
/p:ExcludeObjectType=PartitionFunctions /p:ExcludeObjectType=PartitionSchemes /p:IgnoreIndexOptions=True  /p:IgnoreTableOptions=True /p:DatabaseLockTimeout=60 /p:CommandTimeout=150 
/p:DropObjectsNotInSource=True /p:DropStatisticsNotInSource=False /p:IgnoreObjectPlacementOnPartitionScheme=true 
/p:IgnorePartitionSchemes=true /p:IgnoreTablePartitionOptions=true /p:ScriptDatabaseOptions=False /p:DropPermissionsNotInSource=False /p:DropRoleMembersNotInSource=False 
/p:AdditionalDeploymentContributors=AgileSqlClub.DeploymentFilterContributor /p:AdditionalDeploymentContributorArguments='SqlPackageFilter0=IgnoreType(Statistics);SqlPackageFilter1=IgnoreSchema(tmp)' 
/p:BlockOnPossibleDataLoss=False /p:IncludeCompositeObjects=True /Profile:'post_trade\post_trade_test\Publish\post_trade_test.publish.xml'
</pre>

Tail of `SqlPackage Publish` log with `/d` switch:
<details><summary>Details</summary>
<pre>
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation started (name, details): Top Level Populator,Sql150ExternalTablePopulator
Microsoft.Data.Tools.Diagnostics.Tracer Information: 0 : 2024-02-28T19:04:28 : Populating child populator : Sql150ExternalTableColumnPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 1 : 2024-02-28T19:04:28 : Executing populator:
Sql150ExternalTableColumnPopulator

Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation started (name, details): Child Populator,Sql150ExternalTableColumnPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation ended (name, details, elapsed in ms): Child Populator,Sql150ExternalTableColumnPopulator,0
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation ended (name, details, elapsed in ms): Top Level Populator,Sql150ExternalTablePopulator,0
Microsoft.Data.Tools.Diagnostics.Tracer Information: 0 : 2024-02-28T19:04:28 : Populating populator : SqlSysCommentObjectDependencyPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 1 : 2024-02-28T19:04:28 : Executing populator:
SqlSysCommentObjectDependencyPopulator

Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation started (name, details): Top Level Populator,SqlSysCommentObjectDependencyPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-02-28T19:04:28 : Perf: Operation ended (name, details, elapsed in ms): Top Level Populator,SqlSysCommentObjectDependencyPopulator,21
Microsoft.Data.Tools.Diagnostics.Tracer Warning: 0 : 2024-02-28T19:04:28 : Reverse Engineering operation cancelled. Exception cannot be retried.
 Exception: Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request)
Microsoft.Data.Tools.Diagnostics.Tracer Error: 0 : 2024-02-28T19:04:28 : Unexpected exception in ExecutePopulators().  In a retail build this would go in the ErrorManager rather than being allowed to crash VS.  This should never happen - known errors should be explicitly handled.  Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request) Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request)
Microsoft.Data.Tools.Diagnostics.Tracer Information: 25 : 2024-02-28T19:04:28 : Perf: Operation ended (name, details, elapsed in ms): PopulateAll,,86995
Microsoft.Data.Tools.Diagnostics.Tracer Error: 1 : 2024-02-28T19:04:28 : Error detected when reverse engineering the database. Severity:'Error' Prefix:'' Error Code:'0' Message:An error occurred while attempting to reverse engineer elements of type Microsoft.Data.Tools.Schema.Sql.SchemaModel.ISqlModelElement: Identifier part cannot be null.
Microsoft.Data.Tools.Diagnostics.Tracer Error: 19 : 2024-02-28T19:04:28 : Microsoft.SqlServer.Dac.DacServicesException: An error occurred during deployment plan generation. Deployment cannot continue.
Error SQL72018: ISqlModelElement could not be imported but one or more of these objects exist in your source.
 ---> Microsoft.Data.Tools.Schema.Sql.Deployment.DeploymentFailedException: Errors occurred while modeling the target database.  Deployment can not continue.
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.ThrowIfErrors(String message, ErrorManager errors, Object category)
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeploymentEndpointServer.OnLoad(ErrorManager errors, DeploymentEngineContext context)
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.PrepareModels()
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.InitializePlanGeneratator()
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.CreateController(Action`1 msgHandler)
   at Microsoft.SqlServer.Dac.DacServices.CreateController(SqlDeployment deploymentEngine, ErrorManager errorManager)
   --- End of inner exception stack trace ---
Microsoft.Data.Tools.Diagnostics.Tracer Information: 25 : 2024-02-28T19:04:28 : Perf: Operation ended (name, details, elapsed in ms): Initializing deployment,,91748
Initializing deployment (Failed)
*** An error occurred during deployment plan generation. Deployment cannot continue.
Error SQL72018: ISqlModelElement could not be imported but one or more of these objects exist in your source.

Errors occurred while modeling the target database.  Deployment can not continue.
Microsoft.Data.Tools.Diagnostics.Tracer Information: 18 : 2024-02-28T19:04:28 : Time elapsed: 0:01:32.16.
Time elapsed 0:01:32.16
Microsoft.Data.Tools.Diagnostics.Tracer Information: 0 : 2024-02-28T19:04:28 : SqlPackage completed
</pre>
</details> 

there is about 1.5 million lines of log above it.

Let me know if you need additional details.

**Did this occur in prior versions?  If not - which version(s) did it work in?**
on 162.0.52.1 we had the same issue

(DacFx/SqlPackage/SSMS/Azure Data Studio)

**Comment by namangupta211** [comment]
Hi @IVNSTN, Thanks for highlighting the issue. Can you please share more details or a sample dacpac to help us reproduce the issue?

**Comment by IVNSTN** [comment]
I'd be happy to give you a minimal repro but, as I already mentioned, I have no idea what to look for. The DB is a real thing used at prod. The error is reproduced only in certain circumstances so I'd have to share both full dacpac and DB dump which is impossible.

If you @namangupta211 can give me a hint what to look for in huge publish log I think I could give you some other details. I don't see even what object type it is trying to process.

**Comment by llali** [comment]
@IVNSTN this kind of error usually happens when using * instead of specifying column names in queries.

**Comment by IVNSTN** [comment]
@llali sorry for the delay. Here is the stacktrace from the log generated with /df argument:

```

Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-06-19T12:32:20 : Perf: Operation started (name, details): Child Populator,Sql150ExternalTableColumnPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-06-19T12:32:20 : Perf: Operation ended (name, details, elapsed in ms): Child Populator,Sql150ExternalTableColumnPopulator,0
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-06-19T12:32:20 : Perf: Operation ended (name, details, elapsed in ms): Top Level Populator,Sql150ExternalTablePopulator,0
Microsoft.Data.Tools.Diagnostics.Tracer Information: 0 : 2024-06-19T12:32:20 : Populating populator : SqlSysCommentObjectDependencyPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 1 : 2024-06-19T12:32:20 : Executing populator:
SqlSysCommentObjectDependencyPopulator

Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-06-19T12:32:20 : Perf: Operation started (name, details): Top Level Populator,SqlSysCommentObjectDependencyPopulator
Microsoft.Data.Tools.Diagnostics.Tracer Verbose: 25 : 2024-06-19T12:32:20 : Perf: Operation ended (name, details, elapsed in ms): Top Level Populator,SqlSysCommentObjectDependencyPopulator,20
Microsoft.Data.Tools.Diagnostics.Tracer Warning: 0 : 2024-06-19T12:32:20 : Reverse Engineering operation cancelled. Exception cannot be retried.
 Exception: Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request)
Microsoft.Data.Tools.Diagnostics.Tracer Error: 0 : 2024-06-19T12:32:20 : Unexpected exception in ExecutePopulators().  In a retail build this would go in the ErrorManager rather than being allowed to crash VS.  This should never happen - known errors should be explicitly handled.  Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request) Microsoft.Data.Tools.Schema.SchemaModel.ModelException: Identifier part cannot be null.
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CheckIdentifierPartsLength(ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.SchemaModel.ModelStore.CreateIdentifierImpl(Int32 externalPartsId, ICollection`1 parts)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.GetReferencedElement(Type& referencedElementType, ModelIdentifier& referencedModelId)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.SqlSysCommentObjectDependencyPopulator.CreateElement(ReverseEngineerOption option, EventArgs args)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.ReverseEngineerPopulators.TopLevelElementPopulator`1.Populate(SqlReverseEngineerRequest request, OrdinalSqlDataReader reader, ReverseEngineerOption option)
   at Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlReverseEngineerImpl.ExecutePopulators(Tuple`2 connectionTuple, IList`1 populators, Int32 totalPopulatorsCount, Int32 startIndex, Boolean progressAlreadyUpdated, ReverseEngineerOption option, SqlReverseEngineerRequest request)
Microsoft.Data.Tools.Diagnostics.Tracer Information: 25 : 2024-06-19T12:32:20 : Perf: Operation ended (name, details, elapsed in ms): PopulateAll,,11731
Microsoft.Data.Tools.Diagnostics.Tracer Error: 1 : 2024-06-19T12:32:20 : Error detected when reverse engineering the database. Severity:'Error' Prefix:'' Error Code:'0' Message:An error occurred while attempting to reverse engineer elements of type Microsoft.Data.Tools.Schema.Sql.SchemaModel.ISqlModelElement: Identifier part cannot be null.
Microsoft.Data.Tools.Diagnostics.Tracer Error: 19 : 2024-06-19T12:32:20 : Microsoft.SqlServer.Dac.DacServicesException: An error occurred during deployment plan generation. Deployment cannot continue.
Error SQL72018: ISqlModelElement could not be imported but one or more of these objects exist in your source.
 ---> Microsoft.Data.Tools.Schema.Sql.Deployment.DeploymentFailedException: Errors occurred while modeling the target database.  Deployment can not continue.
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.ThrowIfErrors(String message, ErrorManager errors, Object category)
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeploymentEndpointServer.OnLoad(ErrorManager errors, DeploymentEngineContext context)
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.PrepareModels()
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.InitializePlanGeneratator()
   at Microsoft.Data.Tools.Schema.Sql.Deployment.SqlDeployment.CreateController(Action`1 msgHandler)
   at Microsoft.SqlServer.Dac.DacServices.CreateController(SqlDeployment deploymentEngine, ErrorManager errorManager)
   --- End of inner exception stack trace ---
Microsoft.Data.Tools.Diagnostics.Tracer Information: 25 : 2024-06-19T12:32:20 : Perf: Operation ended (name, details, elapsed in ms): Initializing deployment,,16892
Microsoft.Data.Tools.Diagnostics.Tracer Information: 18 : 2024-06-19T12:32:20 : Time elapsed: 0:00:17.66.
Microsoft.Data.Tools.Diagnostics.Tracer Information: 0 : 2024-06-19T12:32:20 : SqlPackage completed
```

And I sent some additional details directly to your email.

**Comment by llali** [comment]
@IVNSTN  sorry for the delay, you are right about the error is related to one of the select * in the target db. It would be hard for me to figure out what can cause this. what I can do maybe is to add some more information in the log so we know atleast which element is causing the issue.

**Comment by IVNSTN** [comment]
Actually it was your assumption about select *, I tried to check if this can be the reason and it looks like it's not. I sent a bigger log part to your email mentioned in github profile. In full log there are about 1.5 million lines like "resolve this, resolve that". The error occurs after the first time SqlSysCommentObjectDependencyPopulator appears with no additional info. Please check your email.

**Comment by llali** [comment]
@IVNSTN yes I checked the logs but it doesn't have information about which element is causing this. My guess was that "Select * is causing it but I can be wrong. I'm adding more details to the log so using the next release you can see which element is causing this. The other thing you can do is to run the following query in your target database `SELECT
    CASE
        WHEN [d].[referencing_class] = 1 AND NOT [TableType].[user_type_id] IS NULL THEN
            [TableType].[user_type_id]
        ELSE
            [d].[referencing_id] 
        END AS [ReferencingId],
    [o].[type] AS [ReferencingType],
    CASE
        WHEN [d].[referencing_class] = 1 AND NOT [TableType].[user_type_id] IS NULL THEN
            [TableType].[name]
        ELSE [o].[name]
    END AS [ReferencingName],
    [c].[name] AS [ReferencingColumnName],
    CASE
        WHEN [d].[referencing_class] = 1 AND NOT [TableType].[user_type_id] IS NULL THEN
            SCHEMA_NAME([TableType].[schema_id])
        WHEN [d].[referencing_class] = 1 AND NOT [ParentTableType].[user_type_id] IS NULL THEN
            SCHEMA_NAME([ParentTableType].[schema_id])
        ELSE [o].[ReferencingSchemaName]
    END AS [ReferencingSchemaName],
    [d].[referencing_minor_id] AS [ReferencingMinorId],
    CASE 
        WHEN  NOT [ReferencedTableType].[user_type_id] IS NULL THEN [ReferencedTableType].[user_type_id]
        ELSE [d].[referenced_id] 
    END AS [ReferencedId],
    CASE
            WHEN [ReferencedTableType].[user_type_id] IS NULL AND [d].[referenced_schema_name] IS NOT NULL THEN [d].[referenced_schema_name]
            WHEN [d].[referenced_class] = 6 THEN SCHEMA_NAME([udt].[schema_id])
            WHEN d.referenced_class = 10 THEN SCHEMA_NAME([xsc].[schema_id])
            WHEN d.referenced_class = 21 THEN NULL
            WHEN NOT [ReferencedTableType].[user_type_id] IS NULL THEN SCHEMA_NAME([ReferencedTableType].[schema_id])
            WHEN [d].[is_caller_dependent] = 1 AND [d].[referenced_class] = 1 AND [d].[referenced_schema_name] IS NULL THEN (SELECT TOP 1 SCHEMA_NAME(schema_id)  FROM [sys].[objects] AS [o] WITH (NOLOCK) WHERE [o].[name] = [d].[referenced_entity_name])
            ELSE OBJECT_SCHEMA_NAME([d].[referenced_id])
    END AS [ReferencedSchemaName],
    CASE
        WHEN NOT [ReferencedTableType].[user_type_id] IS NULL THEN [ReferencedTableType].[name]
        ELSE [d].[referenced_entity_name] 
    END AS ReferencedName,
    [c2].[name] AS [ReferencedColumnName],
    [d].[referenced_minor_id] AS [ReferencedMinorId],
    CASE [t].[is_filetable] WHEN 1 THEN N'UF' ELSE [o2].[type] END AS [ReferencedType],
    [d].[referenced_class] AS [ReferencedClass],
    [d].[is_caller_dependent] AS [IsCallerDependent],
    [d].[is_ambiguous] AS [IsAmbiguous],
    [np].[procedure_number] AS [ProcedureNumber],
    CAST(ISNULL([parent].[object_id],  0) AS BIT) AS [ReferencingParentIsTableType],
    CASE 
        WHEN [d].[referencing_class]= 1 AND NOT [parent].[object_id] IS NULL THEN N'TT'
        ELSE [o].[type]
    END AS [FilterReferencingType],
    CASE 
        WHEN NOT [ParentTableType].[user_type_id] IS NULL THEN [ParentTableType].[name]
        WHEN [d].[referencing_class] = 1 AND NOT [TableType].[user_type_id] IS NULL THEN
            [TableType].[name]
        ELSE [o].[name]
    END AS [FilterReferencingName],
    CASE
        WHEN NOT [ParentTableType].[user_type_id] IS NULL THEN [ParentTableType].[type_table_object_id]
        ELSE [o].[parent_object_id] 
    END AS [ParentId],
    CASE 
        WHEN [ParentTableType].[user_type_id] IS NOT NULL THEN [ParentTableType].[schema_id]
        ELSE [referencingParent].[schema_id]
    END AS [ParentSchemaId],
    CASE
        WHEN [o].[type] = N'TT' THEN 0
        WHEN [ParentTableType].[user_type_id] IS NOT NULL THEN 0
        ELSE ISNULL([o].[is_ms_shipped], 0) 
    END AS [IsMsShipped],
    [t].[is_dropped_ledger_table] AS [ReferencedDroppedLedgerTable],
    [v].[ledger_view_type] AS [ReferencingLedgerViewType]
    FROM [sys].[sql_expression_dependencies] AS [d] WITH (NOLOCK)
    LEFT JOIN ( 
        SELECT  SCHEMA_NAME([schema_id]) AS [ReferencingSchemaName], [name], [object_id] , 1 AS [parent_class] , [type], [parent_object_id], [is_ms_shipped] FROM [sys].[objects] WITH (NOLOCK)
        UNION
        SELECT [name] AS [ReferencingSchemaName], NULL AS [name], [object_id], [parent_class], N'DT',  0 AS [parent_object_id], [is_ms_shipped] FROM [sys].[triggers] WITH (NOLOCK) where [parent_class] = 0) 
        AS [o] ON [d].[referencing_class] in (1, 12) AND [o].[object_id] = [d].[referencing_id]
    LEFT JOIN [sys].[columns] AS [c] WITH (NOLOCK) ON [d].[referencing_class] = 1 AND [o].[object_id] = [c].[object_id] AND [d].[referencing_minor_id] = [c].[column_id]    
    LEFT JOIN [sys].[objects] AS [referencingParent] WITH (NOLOCK) ON [o].[parent_object_id] = [referencingParent].[object_id]
    LEFT JOIN [sys].[objects] AS [o2] WITH (NOLOCK) ON [d].[referenced_class] = 1 AND [o2].[object_id] = [d].[referenced_id]
    LEFT JOIN [sys].[columns] AS [c2] WITH (NOLOCK) ON [d].[referenced_class] = 1 AND [d].[referenced_id] = [c2].[object_id] AND [d].[referenced_minor_id] = [c2].[column_id]
    LEFT JOIN [sys].[types] as [udt] WITH (NOLOCK) ON [d].[referenced_class] = 6 AND [udt].[user_type_id] = [d].[referenced_id]
    LEFT JOIN [sys].[xml_schema_collections] AS [xsc] WITH (NOLOCK) ON  [d].[referenced_class] = 10 AND [xsc].[xml_collection_id] = [d].[referenced_id]
    LEFT JOIN [sys].[objects] as [parent] WITH (NOLOCK) ON [parent].[object_id] = [o].[parent_object_id] AND [parent].[type] = N'TT'
    LEFT JOIN [sys].[tables] AS [t] WITH (NOLOCK) ON [o2].[object_id] = [t].[object_id]
    LEFT JOIN [sys].[table_types] AS [ParentTableType] WITH (NOLOCK) ON [ParentTableType].[type_table_object_id] = [parent].[object_id]
    LEFT JOIN [sys].[table_types] AS [TableType] WITH (NOLOCK)ON [TableType].[type_table_object_id] = [o].[object_id]
    LEFT JOIN [sys].[table_types] AS [ReferencedTableType] WITH (NOLOCK) ON [ReferencedTableType].[type_table_object_id] = [o2].[object_id]
    --{0}
    LEFT JOIN
    (SELECT [object_id] , 1 AS [procedure_number] FROM [sys].[procedures] WITH (NOLOCK)
        union 
     SELECT [object_id], [procedure_number] FROM [sys].[numbered_procedures] WITH (NOLOCK)) AS [np] ON  [d].[referenced_id] = [np].[object_id] AND [d].[referenced_class] = 1 
    LEFT JOIN [sys].[views] as [v] WITH (NOLOCK) ON [d].[referencing_id] = [v].[object_id]
    WHERE ([o].[ReferencingSchemaName] <> N'cdc' OR [o].[parent_class] <> 1) AND [o].[type] <> N'PG' AND [d].[referenced_database_name] IS NULL AND [d].[referenced_server_name] IS NULL AND [d].[is_ambiguous] = 0 AND  ([d].[is_caller_dependent] = 0 OR 1 = (SELECT COUNT(1) FROM [sys].[objects] WITH (NOLOCK) WHERE [name] = [d].[referenced_entity_name])) AND [c2].[graph_type] IS NULL ;
`


This will give you some info about which element is having a problem. you can check the ReferencedName column for any null value and if there's any, then check the syntax of the element mentioned in ReferencingName column

**Comment by IVNSTN** [comment]
@llali thanks for the query, it did locate some issues in the DB clone. Note, there is garbage `--{0}` in the query above which breaks it.

`ReferencedName` is NULL in several procs and the only common thing in those procs is a reference to the same scalar UDF. And this **function is** totally **missing** in the brand new DB clone. There are other functions a) in the same `dbo` schema b) natively compiled like this one and regular functions c) with begin-atomic block and without it - and all of them are fine in DB clone, none other function is missing.

Currently I'm trying to find out if there is a piece of code in the lost function which can cause the issue. There are no UDTs reverenced, or `SELECT *`.

**Comment by IVNSTN** [comment]
@llali  This is the only function in the database with a reference to memory-optimized temporal/system-versioned table.
It appears that no memory-optimized temporal/system-versioned tables get cloned during DBCC CLONEDATABASE command execution. All such tables (along with their history tables) are missing in clone. The function requiring one of such tables is missing as well.

When I clone not-in-memory database with not-in-memory tables, all the temporal/system-versioned tables with their history tables persist in created clone.

Microsoft SQL Server 2019 (RTM-CU28) (KB5039747) - 15.0.4385.2 (X64)   Jul 25 2024 21:32:40
Windows Server 2016 Standard 10.0 <X64> Build 14393

compatibility_level = 150

**Comment by llali** [comment]
@IVNSTN sorry for the delay. can you please confirm if you add the missing tables and UDTs if you still get the same error from dacfx? so to be able to repro the error, should I create a memory optimized table with UDT columns but UTD types are missing?

**Comment by IVNSTN** [comment]
> There are no UDTs reverenced, or SELECT *.

Sorry if I don't get the point but UDTs are not related to the issue. There are no UDTs mentioned in the system-versioned table script above.

**Comment by llali** [comment]
@IVNSTN sorry I'm just trying to figure out the sample database to repro the bug in dacfx. cloning the database using "DBCC CLONEDATABASE command" and if it has any bug is not related to DacFx operations. we should still notify SQL Server team to report the bug. but when it creates the database, it is in a state which now causes a bug in DacFx. either with missing temporal tables or UTDs ... so that is what I'm trying to fix. If dacfx operation is running on a database with missing objects, it should not throw unhandled exception. I would really appreciate if you can provide me with a small sample to create my database to be able to repro this bug in dacfx.

**Comment by IVNSTN** [comment]
@llali here is a minimal sqlproj with instruction in readme file
[isqlmodelelement_repro.zip](https://github.com/user-attachments/files/17631183/isqlmodelelement_repro.zip)

**Comment by llali** [comment]
@IVNSTN thanks for the detailed info. I was able to repro and fix the bug inside DacFx. The fix will be included in 162.5 release.
Also, about the clone command, looks like it will be deprecated soon to copy databases, DacFx is one of the recommended options.
https://www.microsoft.com/en-us/sql-server/blog/2024/03/25/announcing-the-discontinuation-of-using-dbcc-clonedatabase-command-in-production-deployments/#:~:text=DBCC%20CLONEDATABASE%20command%20generates%20a%20schema-only%20clone%20or,as%20a%20new%20database%20in%20a%20production%20environment.

**Comment by abhikum** [comment]
This issue was fixed in 162.5.57 https://learn.microsoft.com/en-us/sql/tools/sqlpackage/release-notes-sqlpackage?view=sql-server-ver16#162557-sqlpackage
